### PR TITLE
Implement simple RGB driver via digitalWrite; solving #6783

### LINKED ISSUE #6783: Implement LED_BUILTIN constant + digitalWrite() overload for RGB LED
### Related area

RGB LED, Blink example

### Hardware specification

C3, S3, dev boards with RGB LED

### Is your feature request related to a problem?

No issue found

### Describe the solution you'd like

Implement `LED_BUILTIN` constant for dev boards with RGB LED. Where not applicable print a helpful error message - for example, `This board does not have builtin LED` instead of default `error: 'LED_BUILTIN' was not declared in this scope`
For boards with RGB LED overload function `void digitalWrite(uint8_t pin, uint8_t val)` to activate the RGB LED via the appropriate driver.
Sample use of overloaded function:
```
digitalWrite(LED_BUILTIN, HIGH); // RGB turns full white
digitalWrite(LED_BUILTIN, LOW); // RGB turns off

// Nice to have (1) - ability to control each channel
digitalWrite(RED_LED_BUILTIN, HIGH); // Red channel full brightness
digitalWrite(GREEN_LED_BUILTIN, HIGH); // Green channel full brightness
digitalWrite(BLUE_LED_BUILTIN, HIGH); // Blue channel full brightness
// Similar for LOW -> channel is off

// Nice to have (2) - ability to control brightness
digitalWrite(RED_LED_BUILTIN, 128); // Red channel turns on to 50% brightness
```

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### I have checked existing list of Feature requests and the Contribution Guide

- [X] I confirm I have checked existing list of Feature requests and Contribution Guide.

### PR DESCRIPTION
**Pull request by PilnyTomas**
## Description of Change
This solution enables users to use Blink example with RGB LED on board.
When using function `digitalWrite` with constant `LED_BUILTIN` as pin parameter an RGB LED driver in the background will set white / off.
Added new example BlinkRGB extending the original Blink example with a demonstration of separate RGB channel setup.

## Tests scenarios
Tested on
S3 DevkitC
C3 DevkitC
S2 Kaluga

## Related links
Closes #6783

### CODE REVIEW

File: cores/esp32/esp32-hal-gpio.c [hunk 1]
```diff
@@ -91,6 +91,13 @@ static InterruptHandle_t __pinInterruptHandlers[SOC_GPIO_PIN_COUNT] = {0,};
 
 extern void ARDUINO_ISR_ATTR __pinMode(uint8_t pin, uint8_t mode)
 {
+#ifdef BOARD_HAS_NEOPIXEL
```
Comment (igrr) [inline cores/esp32/esp32-hal-gpio.c:94]: Could we consider moving the board-specific code to the `variants` folder?

I'm worried that the more similar features we add for specific boards, the more complex the core code will become. It would be better to handle this in the board variant folder. It would also demonstrate authors of 3rd party board packages how to add such custom functionality.

To avoid duplication of code, you can still define `RGBLedWrite` somewhere in the core, but preferably in a separate source file (esp32-hal-rgb-led.c?)

Comment (PilnyTomas) [reply cores/esp32/esp32-hal-gpio.c:94]: I'm a little bit confused about what should go where.
`pinMode` -> `variants/`  ... new file?
`RGBLedWrite` -> `esp32-hal-rgb-led.c` ?

Comment (PilnyTomas) [reply cores/esp32/esp32-hal-gpio.c]: Should be solved in https://github.com/espressif/arduino-esp32/pull/6808/commits/583e0301427b6e98a87c3fa94fa100d2fc19cb7f

File: cores/esp32/esp32-hal-rgb-led.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+#endif
+
+#include "esp32-hal.h"
+/*
+#include "soc/soc_caps.h"
+#include "pins_arduino.h"
+*/
+
+//#ifdef BOARD_HAS_NEOPIXEL
+  void RGBLedWrite(uint8_t pin, uint8_t red_val, uint8_t green_val, uint8_t blue_val);
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-rgb-led.h:15]: please cleanup the commented includes and ifdefs. Arduino starts function names with lower case. I would suggest that you rename the function to `neopixelWrite`. `rgbLed` is very general and should be attributed to actual RGB Led with separate leads for each color (and controlled through PWM).

File: cores/esp32/esp32-hal-rgb-led.c [hunk 3]
```diff
@@ -0,0 +1,51 @@
+#include "esp32-hal-rgb-led.h"
+
+#ifdef BOARD_HAS_NEOPIXEL
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-rgb-led.c:3]: seems you have the guard here but it is commented in the header?

File: cores/esp32/esp32-hal-rgb-led.c [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+void RGBLedWrite(uint8_t pin, uint8_t red_val, uint8_t green_val, uint8_t blue_val){
+  log_d("RGB: %d %d %d", red_val, green_val, blue_val);
+  rmt_data_t led_data[24];
+  static rmt_obj_t* rmt_send = NULL;
+  static bool initialized = false;
+
+  uint8_t _pin;
+  if(pin == LED_BUILTIN){
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-rgb-led.c:12]: this should be evaluated only if the board has neopixel, else LED_BUILTIN would be lower than SOC_GPIO_PIN_COUNT

File: cores/esp32/esp32-hal-rgb-led.c [hunk 5]
```diff
@@ -0,0 +1,51 @@
+#include "esp32-hal-rgb-led.h"
+
+#ifdef BOARD_HAS_NEOPIXEL
+
+void RGBLedWrite(uint8_t pin, uint8_t red_val, uint8_t green_val, uint8_t blue_val){
+  log_d("RGB: %d %d %d", red_val, green_val, blue_val);
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-rgb-led.c:6]: such logs should be with level "Verbose" instead. Blinking led will print a lot of things :)

File: cores/esp32/esp32-hal-rgb-led.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "esp32-hal.h"
+/*
+#include "soc/soc_caps.h"
+#include "pins_arduino.h"
+*/
+
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-rgb-led.h:13]: You should have safeguard against LED_BRIGHTNESS not being defined:
```cpp
#ifndef LED_BRIGHTNESS
#define LED_BRIGHTNESS 64
#endif
```

File: variants/esp32s2/variant.cpp [hunk 7]
```diff
@@ -0,0 +1,13 @@
+#include "esp32-hal-rgb-led.h"
+
+extern void ARDUINO_ISR_ATTR __pinMode(uint8_t pin, uint8_t mode)
+{
+  log_d("foo");
```
Comment (me-no-dev) [inline variants/esp32s2/variant.cpp:5]: forgotten?

File: variants/esp32c3/variant.cpp [hunk 8]
```diff
@@ -0,0 +1,13 @@
+#include "esp32-hal-rgb-led.h"
+
+extern void ARDUINO_ISR_ATTR __pinMode(uint8_t pin, uint8_t mode)
+{
+  log_d("foo");
```
Comment (me-no-dev) [inline variants/esp32c3/variant.cpp:5]: forgotten?

File: variants/esp32c3/variant.cpp [hunk 9]
```diff
@@ -0,0 +1,13 @@
+#include "esp32-hal-rgb-led.h"
+
+extern void ARDUINO_ISR_ATTR __pinMode(uint8_t pin, uint8_t mode)
```
Comment (me-no-dev) [inline variants/esp32c3/variant.cpp:3]: why is this necessary? pinMode in the core should just return if pin is not valid.

File: cores/esp32/esp32-hal-rgb-led.c (hunk 5 above)
Comment (PilnyTomas) [reply cores/esp32/esp32-hal-rgb-led.c:6]: this was also a forgotten debug output. I will remove it in the next commit.

File: cores/esp32/esp32-hal-rgb-led.c (hunk 4 above)
Comment (PilnyTomas) [reply cores/esp32/esp32-hal-rgb-led.c:12]: The entire function was safeguarded `#ifdef BOARD_HAS_NEOPIXEL`
But I will remove safeguard for the entire function a wrap only this.

File: variants/esp32c3/variant.cpp (hunk 9 above)
Comment (PilnyTomas) [reply variants/esp32c3/variant.cpp:3]: I was trying to implement this https://github.com/espressif/arduino-esp32/pull/6808#pullrequestreview-986096611

Comment (me-no-dev) [reply variants/esp32c3/variant.cpp:3]: I understand, I just question the necessity to have such code in the variant just to get something as common as neopixel on the board working.
@igrr any comments?

Something like this in the core's `__pinMode` would skip execution:
```cpp
if(pin > SOC_GPIO_PIN_COUNT && pin == LED_BUILTIN){
    return; // or pin -= SOC_GPIO_PIN_COUNT; to allow setting the mode.
}
```

Comment (me-no-dev) [reply variants/esp32c3/variant.cpp:3]: @PilnyTomas this is still outstanding. Requiring variant.cpp to blink a led is not desired situation.

Comment (PilnyTomas) [reply variants/esp32c3/variant.cpp:3]: @igrr Could you please give us more info about why should we have board-specific code in the variants folder instead of having in the common file?
Should I keep it like this, or revert back to the common file?

Comment (igrr) [reply variants/esp32c3/variant.cpp:3]: I was thinking that this is behavior we want only on specific boards which have a WS2812B, and where we'd like to provide such emulation.

Makers of other boards should be able implement some different logic, or opt out of such emulation. In the latter case, the code for emulating an LED ideally shouldn't be compiled or linked.

Comment (me-no-dev) [reply variants/esp32c3/variant.cpp:3]: if you do not have that LED defined, code will not be compiled anyway. requiring all boards with RGB leds to include a variant cpp for two lines is a bit too much IMO. There are plenty boards with RGB leds

Comment (igrr) [reply variants/esp32c3/variant.cpp:3]: Although, given that you are introducing a BOARD_HAS_NEOPIXEL macro, maybe we can indeed move this functionality into the core.
